### ISSUE TITLE
Closure iterator finishing prematurely

### ISSUE THREAD

**Issue by dawkot** [nim-lang/Nim#11042]
```nim
iterator foo: int =
  for x in 1..2:
    echo "foo: ", x
    for y in 1..2:
      discard

for x in foo(): discard

let bar = iterator: int =
  for x in 1..2:
    echo "bar: ", x
    for y in 1..2:
      discard

for x in bar(): discard
```
output:
> foo: 1
> foo: 2
> bar: 1

**Comment by jyapayne** [comment]
Something is definitely weird here. In addition to the behavior above, if I change your code to:

```nim
let bar = iterator: int =
  for x in 1..9:
    echo "bar: ", x
    for y in 1..2:
      discard
```

(changed the 2 to a 9 in the first loop), I get **an infinite loop**(!!).

**Comment by jyapayne** [comment]
I should note that this only happens on top level declarations. If you wrap everything in a proc it seems to work fine.

**Comment by Menduist** [comment]
Now works correctly on devel (got fixed somewhere between 1.4 & 1.6.6) (including the second snippet)
cc @xflywind you can probably close this

**Comment by ringabout** [comment]
@Menduist thanks for reporting it! I will close it soon.
